engiopt/pixel_cnn_pp_2d/pixel_cnn_pp_2d.py

```diff
@@ -0,0 +1,687 @@
+        xs = list(x.shape) # true image (i.e. labels)
+        ls = list(l.shape) # predicted distribution
+
+        # unpacking the params of the mixture of logistics
+        nr_mix = int(ls[-1] / 3)
+        logit_probs = l[:,:,:,:nr_mix] # mixture probabilities
+        l = l[:, :, :, nr_mix:].contiguous().view([*xs, nr_mix * 2]) # 2 for mean, scale
+        means = l[:,:,:,:,:nr_mix]
+        log_scales = th.clamp(l[:, :, :, :, nr_mix:2 * nr_mix], min=-7.)
+        x = x.contiguous()
+        zeros = th.zeros([*xs, nr_mix], device=x.device)
+        x = x.unsqueeze(-1) + zeros
+
+        centered_x = x - means
+        inv_stdv = th.exp(-log_scales)
+        plus_in = inv_stdv * (centered_x + 1./255.)
+        cdf_plus = f.sigmoid(plus_in)
+        min_in = inv_stdv * (centered_x - 1./255.)
+        cdf_min = f.sigmoid(min_in)
+        # log probability for edge case of 0
+        log_cdf_plus = plus_in - f.softplus(plus_in)
+        # log probability for edge case of 255
+        log_one_minus_cdf_min = -f.softplus(min_in)
+        # probability for all other cases
+        cdf_delta = cdf_plus - cdf_min
+        # log probability in the center of the bin, to be used in extreme cases
+        mid_in = inv_stdv * centered_x
+        log_pdf_mid = mid_in - log_scales - 2.*f.softplus(mid_in)
+
+        # select the right output: left edge case, right edge case, normal case, extremely low prob case (doesn't actually happen here)
+
+        # this is what is really done, but using the robust version below for extreme cases
+        # log_probs = th.where(x < -0.999, log_cdf_plus, th.where(x > 0.999, log_one_minus_cdf_min, th.log(cdf_delta)))  # noqa: ERA001
+
+        # robust version, that still works if the probability is below 1e-5
+        # approximation used based on the assumption that the log-density is constant in the bin of the observed sub-pixel value
+        log_probs = th.where(x < -0.999, log_cdf_plus, th.where(x > 0.999, log_one_minus_cdf_min, th.where(cdf_delta > 1e-5, th.log(th.clamp(cdf_delta, min=1e-12)), log_pdf_mid - np.log(127.5))))  # noqa: PLR2004
+        log_probs = th.sum(log_probs, dim=3) + log_prob_from_logits(logit_probs)
+        return -th.sum(log_sum_exp(log_probs))
+
+
+
+def to_one_hot(tensor: th.Tensor, n: int, fill_with: float = 1.) -> th.Tensor:
+    """One hot encoding with respect to the last axis."""
+    one_hot = th.zeros((*tensor.size(), n), device=tensor.device)
+    one_hot.scatter_(len(tensor.size()), tensor.unsqueeze(-1), fill_with)
+    return one_hot
+
+
+
+def sample_from_discretized_mix_logistic(l: th.Tensor, nr_mix: int) -> th.Tensor:
+    """Sample from a discretized mixture of logistic distributions."""
+    # Tensorflow ordering
+    l = l.permute(0, 2, 3, 1)
+    ls = list(l.shape)
+    xs = [*ls[:-1], 1]
+
+    # unpack parameters
+    logit_probs = l[:, :, :, :nr_mix] # mixture probabilities
+    l = l[:, :, :, nr_mix:].contiguous().view([*xs, nr_mix * 2]) # for mean, scale
```

Comment:
This is very cryptic to me, a comment would help understanding what's going on here.